### PR TITLE
Official PowerShell Package pipeline

### PR DESCRIPTION
**Pull request by adityapatwardhan**
# PR Summary

Migrate the official package pipleine to OneBranch. 

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: .pipelines/templates/insert-nuget-config-azfeed.yml [hunk 1]
```diff
@@ -20,7 +20,7 @@ steps:
 - pwsh: |
     $configPath = "${env:NugetConfigDir}/nuget.config"
     Import-Module ${{ parameters.repoRoot }}/build.psm1 -Force
-    New-NugetConfigFile -NugetFeedUrl $(PowerShellCore_PublicPackages) -UserName $(AzDevOpsFeedUserName) -ClearTextPAT $(AzDevOpsFeedUserName) -FeedName AzDevOpsFeed -Destination "${env:NugetConfigDir}"
+    New-NugetConfigFile -NugetFeedUrl $(PowerShellCore_PublicPackages) -UserName $(AzDevOpsFeedUserName) -ClearTextPAT $(AzDevOpsFeedPAT2) -FeedName AzDevOpsFeed -Destination "${env:NugetConfigDir}"
```
Comment (TravisEz13) [inline .pipelines/templates/insert-nuget-config-azfeed.yml:23]: discuss PAT compliance.

File: .pipelines/PowerShell-Packages-Official.yml [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    value: ${{ parameters.debug }}
+  - name: ENABLE_PRS_DELAYSIGN
+    value: 1
+  - name: ROOT
+    value: $(Build.SourcesDirectory)
+  - name: NUGET_XMLDOC_MODE
+    value: none
+  - name: nugetMultiFeedWarnLevel
+    value: none
+  - name: NugetSecurityAnalysisWarningLevel
```
Comment (TravisEz13) [inline .pipelines/PowerShell-Packages-Official.yml:34]: In theory some of these might not be needed with the restore/build phase model.

Comment (TravisEz13) [reply .pipelines/PowerShell-Packages-Official.yml:34]: Can defer work, Please have Jason track.

File: .pipelines/templates/linux-package-build.yml [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+      Write-Verbose -Verbose "Signed artifacts"
+      Get-ChildItem "$(Pipeline.Workspace)/CoOrdinatedBuildPipeline/${{ parameters.signedDrop }}" -Recurse
+    displayName: 'Capture Downloaded Artifacts'
+    # Diagnostics is not critical it passes every time it runs
+    continueOnError: true
+    env:
+      ob_restore_phase: true # This ensures checkout is done at the beginning of the restore phase
+
+  - pwsh: |
+      $packageType = '$(PackageType)'
```
Comment (TravisEz13) [inline .pipelines/templates/linux-package-build.yml:88]: This is a significant script, should be a file.

not blocking.

File: .pipelines/templates/linux-package-build.yml [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+  - pwsh: |
+      $pkgFilter = '$(PackageFilter)'
+      Write-Verbose -Verbose "pkgFilter: $pkgFilter"
+
+      $pkgPath = Get-ChildItem -Path $(Pipeline.Workspace) -Filter $pkgFilter -Recurse -File | Select-Object -ExpandProperty FullName
+      Write-Verbose -Verbose "pkgPath: $pkgPath"
+      Copy-Item -Path $pkgPath -Destination '$(ob_outputDirectory)' -Force -Verbose
+    displayName: 'Copy artifacts to output directory'
+    env:
+      __DOTNET_RUNTIME_FEED_KEY: $(RUNTIME_SOURCEFEED_KEY)
```
Comment (TravisEz13) [inline .pipelines/templates/linux-package-build.yml:178]: We need to figure out how to get this automatically.

Not blocking

File: .pipelines/templates/nupkg.yml [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+    artifact: drop_linux_sign_linux_fxd
+    displayName: 'Download drop_linux_sign_linux_fxd'
+    env:
+      ob_restore_phase: true # This ensures checkout is done at the beginning of the restore phase
+
+  - download: CoOrdinatedBuildPipeline
+    artifact: drop_linux_sign_linux_fxd_x64_alpine
+    displayName: 'Download drop_linux_sign_linux_fxd_x64_alpine'
+    env:
+      ob_restore_phase: true # This ensures checkout is done at the beginning of the restore phase
```
Comment (TravisEz13) [inline .pipelines/templates/nupkg.yml:74]: comment is wrong on this option in several places

File: tools/packaging/packaging.psm1 [hunk 6]
```diff
@@ -2409,6 +2423,14 @@ function New-ILNugetPackageSource
 
         Write-Log "Copied the built-in modules to contentFiles for the SDK package"
     }
+    else {
+        Write-Verbose -Verbose "Skipping copying the built-in modules and reference assemblies for $FileName"
+    }
+
+    if (-not $PSBoundParameters.ContainsKey("CGManifestPath")) {
+        Write-Verbose -Verbose "CGManifestPath is not provided. Skipping CGManifest creation."
```
Comment (TravisEz13) [inline tools/packaging/packaging.psm1:2431]: discuss SBOM

**Review by TravisEz13 — COMMENTED** [review]
need to discuss

File: tools/packaging/packaging.psm1 (hunk 6 above)
Comment (adityapatwardhan) [reply tools/packaging/packaging.psm1:2431]: Ask Jason to add to the backlog

File: .pipelines/templates/insert-nuget-config-azfeed.yml (hunk 1 above)
Comment (adityapatwardhan) [reply .pipelines/templates/insert-nuget-config-azfeed.yml:23]: Fixed

File: tools/packaging/packaging.psm1 (hunk 6 above)
Comment (adityapatwardhan) [reply tools/packaging/packaging.psm1:2431]: Jason added all items to backlog

File: .pipelines/PowerShell-Packages-Official.yml [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+  - name: skipNugetSecurityAnalysis
+    value: true
+  - name: ReleaseTagVar
+    value: ${{ parameters.ReleaseTagVar }}
+  - name: ob_outputDirectory
+    value: '$(Build.ArtifactStagingDirectory)/ONEBRANCH_ARTIFACT'
+  - name: WindowsContainerImage
+    value: 'onebranch.azurecr.io/windows/ltsc2019/vse2022:latest' # Docker image which is used to build the project https://aka.ms/obpipelines/containers
+  - name: LinuxContainerImage
+    value: onebranch.azurecr.io/linux/ubuntu-2004:latest
```
Comment (TravisEz13) [inline .pipelines/PowerShell-Packages-Official.yml:45]: I didn't notice this, we need to discuss.

**Review by TravisEz13 — APPROVED** [review]

**Review by Moazzem-Hossain-pixel — COMMENTED** [review]
Need support for resolving this issue

**Pull request merged**